### PR TITLE
Adds documentation for `Mappable` and `Bindable` interfaces

### PR DESCRIPTION
**Pull request by thepabloaguilar**
refs #517

### CODE REVIEW

File: docs/pages/interfaces.rst [hunk 1]
```diff
@@ -3,8 +3,101 @@
...
+  >>> # We have a shortcut for each amount of arguments to reduce the boilerplate
+  >>> one_type: Mappable1[int]
+  >>> two_types: Mappable2[int, str]
+  >>> three_type: Mappable3[int, str, bool]
+
+.. note::
+
+  Useful links before you start here:
+
+  * `Functors, Applicatives, And Monads In Pictures <http://adit.io/posts/2013-04-17-functors,_applicatives,_and_monads_in_pictures.html>`_
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:33]: Let's replace "functor", "monad" etc with our own names: mappable, bindable, etc.
See #451

File: docs/pages/interfaces.rst [hunk 2]
```diff
@@ -3,8 +3,101 @@
...
+  * `Understanding Functor and Monad With a Bag of Peanuts <https://medium.com/beingprofessional/understanding-functor-and-monad-with-a-bag-of-peanuts-8fa702b3f69e>`_
+  * `Variance of generic types <https://mypy.readthedocs.io/en/latest/generics.html#variance-of-generic-types>`_
+  * `If you know map, I will teach you monads <https://www.youtube.com/watch?v=cB0vpg9-YMQ>`_
+
+Mappable
+--------
+
+:class:`Mappable <returns.interfaces.mappable.MappableN>` interface help us to
+create our own ``Functor`` like :class:`Result <returns.maybe.Maybe>`.
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:42]: Let's focus on the interface. `Mappable` is something with `.map` method. That's it.

And then, we can explain why this method is helpful in general.

Comment (sobolevn) [reply docs/pages/interfaces.rst:42]: Btw, 
```
:class:`Result <returns.maybe.Maybe>`
```

Should be: 
```
:class:`Maybe <returns.maybe.Maybe>`
```

File: docs/pages/interfaces.rst [hunk 3]
```diff
@@ -3,8 +3,101 @@
...
+.. code:: python
+
+  >>> from typing import Any, Callable, TypeVar
+
+  >>> from returns.interfaces.mappable import Mappable1
+
+  >>> _FunctorValueType = TypeVar('_FunctorValueType', covariant=True)
+  >>> _NewFunctorValueType = TypeVar('_NewFunctorValueType')
+
+  >>> class MyFunctor(Mappable1[_FunctorValueType]):
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:53]: Let's give it a unique name, like `Number`

And we will end up illustarting how one can compose `Number`s and math functions.

File: docs/pages/interfaces.rst [hunk 4]
```diff
@@ -3,8 +3,101 @@
...
+  ...             return False
+  ...         return self._inner_value == other._inner_value
+
+Laws
+~~~~
+
+To make sure your functor implementation is right, you can apply the
+Functors law on it to test.
+
+1. **Identity Law:** When we pass the identity function to the map method,
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:77]: I would suggest to explain **why** do we need this laws in the first place.
Let's show some bad situations when these laws are not respected.

Then, we can show that these laws can save you from yourself. 
Related #457

File: docs/pages/interfaces.rst (hunk 1 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst:33]: `Functors, Applicatives, And Monads In Pictures` is the article's name, can be a little weird change the name to `Mappable, Applicative, Bindable in Pictures`. The user can be confused hahaha

But in the other scenarios, I agree with you!

Comment (sobolevn) [reply docs/pages/interfaces.rst:33]: Oh, sorry. Yes, articles can be named whatever they originally have in their title.

Comment (thepabloaguilar) [reply docs/pages/interfaces.rst:33]: Done!

File: docs/pages/interfaces.rst (hunk 2 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 3 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst [hunk 5]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
 Mappable
 --------
 
-:class:`Mappable <returns.interfaces.mappable.MappableN>` interface help us to
-create our own ``Functor`` like :class:`Result <returns.maybe.Maybe>`.
+Something is considered mappable if we can `map` it using a function, generally
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:41]: A formatting fix, not:

```
`map`
```

but 

```
``map``
```

File: docs/pages/interfaces.rst [hunk 6]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
+`map` is a method that accepts a function. An example in this library is
+:class:`Maybe <returns.maybe.Maybe>`, it has the `map` method.
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:43]:
```
An example in this library is
:class:`Maybe <returns.maybe.Maybe>`, it has the `map` method:

.. code:: python

   >>> # example body

```

File: docs/pages/interfaces.rst [hunk 7]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
 
 .. code:: python
 
   >>> from typing import Any, Callable, TypeVar
 
   >>> from returns.interfaces.mappable import Mappable1
 
-  >>> _FunctorValueType = TypeVar('_FunctorValueType', covariant=True)
-  >>> _NewFunctorValueType = TypeVar('_NewFunctorValueType')
+  >>> _NumberType = TypeVar('_NumberType', covariant=True)
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:55]: We can remove `covariant=True` here, I guess that it is not important, but simplifies the example.

File: docs/pages/interfaces.rst [hunk 8]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
+  >>> _NewNumberType = TypeVar('_NewNumberType')
 
-  >>> class MyFunctor(Mappable1[_FunctorValueType]):
-  ...     _inner_value: _FunctorValueType
+  >>> class Number(Mappable1[_NumberType]):
+  ...     _inner_value: _NumberType
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:59]: We can skip `_inner_value` annotation here.

File: docs/pages/interfaces.rst (hunk 7 above)
Comment (sobolevn) [reply docs/pages/interfaces.rst:55]: Let's also bound `_NumberType` to `Union[int, float, complex]`

File: docs/pages/interfaces.rst [hunk 9]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
   ...
   ...     def map(  # This method is required by Mappable
   ...         self,
-  ...         function: Callable[[_FunctorValueType], _NewFunctorValueType]
-  ...     ) -> 'MyFunctor[_NewFunctorValueType]':
+  ...         function: Callable[[_NumberType], _NewNumberType]
+  ...     ) -> 'Number[_NewNumberType]':
   ...         new_value = function(self._inner_value)
-  ...         return MyFunctor(new_value)
+  ...         return Number(new_value)
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:69]: Let's make a single line here, it is way more ideomatic.

File: docs/pages/interfaces.rst [hunk 10]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
   ...
   ...     def __eq__(self, other: Any) -> bool:  # Required to check the identity law
   ...         if not isinstance(self, type(other)):
   ...             return False
   ...         return self._inner_value == other._inner_value
 
+With our ``Number`` mappable class we can compose easily math functions on it.
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:76]: `with it`

File: docs/pages/interfaces.rst [hunk 11]
```diff
@@ -38,36 +38,53 @@ for example, can have one, two or three possible types. See the example below:
...
+
+.. code:: python
+
+  >>> from math import floor, sqrt
+
+  >>> def my_math_function(number: int) -> int:
+  ...     return number + 1
+
+  >>> number: Number[int] = Number(-82)
+  >>> assert number.map(my_math_function).map(abs).map(sqrt).map(floor) == Number(9)
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:86]: Just two is enough, currently it is hard to follow

File: docs/pages/interfaces.rst (hunk 5 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 6 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 7 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 8 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 9 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 10 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 11 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 4 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst:100]: Do you have some bad situations examples? And a link explain why we need the laws?
Currently I don't have this material!

File: docs/pages/interfaces.rst [hunk 12]
```diff
@@ -38,52 +38,58 @@ for example, can have one, two or three possible types. See the example below:
...
 
 .. code:: python
 
-  >>> from typing import Any, Callable, TypeVar
+  >>> from returns.maybe import Maybe, Some
+
+  >>> def mappable(string: str) -> str:
+  ...     return string + '!'
+
+  >>> maybe_str: Maybe[str] = Maybe.from_value('example')
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:54]: It is better to use the same constructors on the both sides: `Maybe.from_value` or `Some`.

So:

```python
maybe_str: Maybe[str] = Some('example')
assert maybe_str.map(mappable) == Some('example!')
```

File: docs/pages/interfaces.rst [hunk 13]
```diff
@@ -38,52 +38,58 @@ for example, can have one, two or three possible types. See the example below:
...
+
+.. code:: python
+
+  >>> from typing import Any, Callable, TypeVar, Union
 
   >>> from returns.interfaces.mappable import Mappable1
 
-  >>> _NumberType = TypeVar('_NumberType', covariant=True)
-  >>> _NewNumberType = TypeVar('_NewNumberType')
+  >>> _NumberType = TypeVar('_NumberType', bound=Union[int, float, complex])
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:66]: Related https://github.com/dry-python/returns/issues/474

**Comment by sobolevn** [comment]
Writting docs is really hard! 😅

File: docs/pages/interfaces.rst (hunk 12 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: Done!

File: docs/pages/interfaces.rst (hunk 13 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst]: I've removed the `bound` parameter!

**Comment by thepabloaguilar** [comment]
> Writting docs is really hard! 😅

Yes, it's but can help us to improve our explanation skills, maybe! 😄 😄

File: docs/pages/interfaces.rst (hunk 4 above)
Comment (sobolevn) [reply docs/pages/interfaces.rst:104]: I actually don't have an experience with broken monads, but let's think of it right now 😅 
I can only think of `Promise` from `js` that sometimes is very hard to use because of it. Like https://buzzdecafe.github.io/2018/04/10/no-promises-are-not-monads

@astynax can you please help us here? Do you have any bad examples in mind?

Comment (sobolevn) [reply docs/pages/interfaces.rst:104]: I also invite @morozzzko to collaborate!

Comment (sobolevn) [reply docs/pages/interfaces.rst:104]: Here are some example of why it is bad: https://mmhaskell.com/monads/laws

Comment (Morozzzko) [reply docs/pages/interfaces.rst:104]: Well, I can clearly see a confusion here. Mainly because I'm confused.

How do we explain **why** an implementation must respect the rule, if the rule is an important part of the definition and the interface? 

I mean, it's just a part of the abstraction. A convention which we use to make it useful. If we want our mappable thing to be useful, it has to follow some rules. Otherwise it just becomes extremely hard to reason about & difficult to work with. 

Let's say we have a mappable which does not respect the identity law. What does it do then? What does it map to? Why does it do that? What's the design benefit? 

I think it's way better not to focus on the **why** and just mention that those lows _make the abstraction work_. 

Perhaps, the docs may ask to send an email if anybody's got a legitimate case when the law must be broken. Would be extremely educational for all parts

Comment (sobolevn) [reply docs/pages/interfaces.rst:104]: We can also give an example where `fmap id != id`, because we can create an implementation where `fmap id` produces a side-effect

Comment (thepabloaguilar) [reply docs/pages/interfaces.rst:104]: I'll add this code below as an example:
```python
class Number(...):
    def map(self, function: Callable[[_Type], _NewType]) -> 'Number[_NewType]':
        return Number(2 * function(self._inner_value))
```

File: docs/pages/interfaces.rst [hunk 14]
```diff
@@ -3,8 +3,234 @@
...
+the mappable has to be the same, unaltered.
+
+.. code:: python
+
+  >>> from returns.functions import identity
+
+  >>> mappable_number: Number[int] = Number(1)
+  >>> assert mappable_number.map(identity) == Number(1)
+
+2. **Associative Law**: Given two functions, `x` and `y`, calling the map
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:114]:
```
`x`
```

->

```
``x``
```

Comment (sobolevn) [reply docs/pages/interfaces.rst:114]: and the same for every other code block 👍

File: docs/pages/interfaces.rst [hunk 15]
```diff
@@ -3,8 +3,234 @@
...
+.. code:: python
+
+  >>> def add_one(number: int) -> int:
+  ...     return number + 1
+
+  >>> def multiply_by_ten(number: int) -> int:
+  ...     return number * 10
+
+  >>> mappable_number: Number[int] = Number(9)
+  >>> assert mappable_number.map(add_one).map(multiply_by_ten) == mappable_number.map(lambda value: multiply_by_ten(add_one(value)))
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:127]: What do you think, should we use `compose` here?

File: docs/pages/interfaces.rst [hunk 16]
```diff
@@ -3,8 +3,234 @@
...
+  >>> from typing import Any, Callable, TypeVar
+
+  >>> from returns.interfaces.bindable import Bindable1
+  >>> from returns.primitives.hkt import Kind1
+
+  >>> _BagContentType = TypeVar('_BagContentType')
+  >>> _NewBagContentType = TypeVar('_NewBagContentType')
+
+  >>> class Bag(
+  ...     Kind1['Bag', int],
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:149]: `SupportsKind1`

File: docs/pages/interfaces.rst [hunk 17]
```diff
@@ -3,8 +3,234 @@
...
+  >>> from typing import Any, Callable, TypeVar
+
+  >>> from returns.interfaces.mappable import Mappable1
+  >>> from returns.primitives.hkt import Kind1
+
+  >>> _NumberType = TypeVar('_NumberType')
+  >>> _NewNumberType = TypeVar('_NewNumberType')
+
+  >>> class Number(
+  ...     Kind1['Number', _NumberType],
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:71]: `SupportsKind1`

File: docs/pages/interfaces.rst [hunk 18]
```diff
@@ -3,8 +3,234 @@
...
+  ...         return self._inner_value == other._inner_value
+
+  >>> @dataclass
+  ... class Peanuts:
+  ...     quantity: int
+
+  >>> def get_half(peanuts: Peanuts) -> Bag[Peanuts]:
+  ...     return Bag(Peanuts(peanuts.quantity // 2))
+
+  >>> bag_of_peanuts: Bag[Peanuts] = Bag(Peanuts(10))
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:173]: Awesome example! 👏

File: docs/pages/interfaces.rst [hunk 19]
```diff
@@ -3,8 +3,234 @@
...
+  >>> assert bag.bind(minus_one).bind(half) == bag.bind(lambda value: minus_one(value).bind(half))
+
+What's the difference between ``Mappable`` and ``Bindable``?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While Mappable you have to pass a pure function, like:
+
+.. code:: python
+
+  >>> def mappable(string: str) -> str:
```
Comment (sobolevn) [inline docs/pages/interfaces.rst:223]: Let's not name functions `mappable` and `bindable` here. Readers might confuse them with `Mappable` and `Bindable` interfaces. Let's maybe name them `can_be_mapped` and `can_be_bound`?

File: docs/pages/interfaces.rst (hunk 15 above)
Comment (thepabloaguilar) [reply docs/pages/interfaces.rst:127]: I think so, maybe it'll be more clear to understand the `law`!!

**Comment by thepabloaguilar** [comment]
I've decided to split this PR because is too many information to just one PR! (#517)

**Comment by sobolevn** [comment]
We can work from here! Thanks a lot!

You have built a great foundation for the interfaces docs!
